Source/Particles/Resampling/VelocityCoincidenceThinning.cpp

```diff
@@ -0,0 +1,284 @@
+/* Copyright 2024 The WarpX Community
+ *
+ * This file is part of WarpX.
+ *
+ * Authors: Roelof Groenewald (TAE Technologies)
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "VelocityCoincidenceThinning.H"
+
+#include "Particles/Algorithms/KineticEnergy.H"
+#include "Utils/Parser/ParserUtils.H"
+#include "Utils/ParticleUtils.H"
+
+
+VelocityCoincidenceThinning::VelocityCoincidenceThinning (const std::string& species_name)
+{
+    using namespace amrex::literals;
+
+    const amrex::ParmParse pp_species_name(species_name);
+
+    utils::parser::queryWithParser(
+        pp_species_name, "resampling_min_ppc", m_min_ppc
+    );
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
+        m_min_ppc >= 1,
+        "Resampling min_ppc should be greater than or equal to 1"
+    );
+
+    utils::parser::getWithParser(
+        pp_species_name, "resampling_algorithm_delta_ur", m_delta_ur
+    );
+    utils::parser::getWithParser(
+        pp_species_name, "resampling_algorithm_n_theta", m_ntheta
+    );
+    utils::parser::getWithParser(
+        pp_species_name, "resampling_algorithm_n_phi", m_nphi
+    );
+}
+
+void VelocityCoincidenceThinning::operator() (WarpXParIter& pti, const int lev,
+                                   WarpXParticleContainer * const pc) const
+{
+    using namespace amrex::literals;
+
+    auto& ptile = pc->ParticlesAt(lev, pti);
+    auto& soa = ptile.GetStructOfArrays();
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_3D)
+    auto * const AMREX_RESTRICT x = soa.GetRealData(PIdx::x).data();
+#elif defined(WARPX_DIM_RZ)
+    auto * const AMREX_RESTRICT x = soa.GetRealData(PIdx::x).data(); // rename to PIdx::r after PR #4667
+#endif
+#if defined(WARPX_DIM_3D)
+    auto * const AMREX_RESTRICT y = soa.GetRealData(PIdx::y).data();
+#endif
+    auto * const AMREX_RESTRICT z = soa.GetRealData(PIdx::z).data();
+    auto * const AMREX_RESTRICT ux = soa.GetRealData(PIdx::ux).data();
+    auto * const AMREX_RESTRICT uy = soa.GetRealData(PIdx::uy).data();
+    auto * const AMREX_RESTRICT uz = soa.GetRealData(PIdx::uz).data();
+    auto * const AMREX_RESTRICT w = soa.GetRealData(PIdx::w).data();
+    auto * const AMREX_RESTRICT idcpu = soa.GetIdCPUData().data();
+
+    // Using this function means that we must loop over the cells in the ParallelFor.
+    auto bins = ParticleUtils::findParticlesInEachCell(lev, pti, ptile);
+
+    const auto n_cells = static_cast<int>(bins.numBins());
+    auto *const indices = bins.permutationPtr();
+    auto *const cell_offsets = bins.offsetsPtr();
+
+    const auto min_ppc = m_min_ppc;
+
+    const auto mass = pc->getMass();
+
+    // create a GPU vector to hold the momentum cluster index for each particle
+    amrex::Gpu::DeviceVector<int> momentum_bin_number;
```

Comment:
This could be simplified
```suggestion
    amrex::Gpu::DeviceVector<int> momentum_bin_number(bins.numItems());
```